port-net: Extend counter-op mode to 4 players

iamgreaser opened the pull request:
Well, where do we start... This extends counter-op mode to up to 4 local players, and the only reason for it being 4 is because that's the practical limit on the number of local players. Trying to go beyond 4 players for split-screen would require extra design work including making menus even narrower than what's offered in vertical split mode.

But for netplay integration purposes (not implemented yet!), going beyond 4 players should not be (much of) a problem.

At this stage it's quite playable. There are a few subtle cases of jank which most people aren't going to hit, and if you know about some of them then it's best to make sure that you're playing with people who are at least mature enough to not try to ruin the game.

Ref: #383 (extended *co-op* mode + what that would entail, and some elaboration on what that would take for counter-op and why counter-op mode is easier to implement)

----

Here's what's implemented:

- A bunch of logic which determines whether a player is a protagonist or an antagonist has been inverted from "is anti player" to "is not bond player". In practice, this needed to be inverted in about 20 different places.
- Instead of selecting the counter-operative player from a list of 2 players, you select the main player from a list of 2 through 4 players.
- 3-way and 4-way split screen now uses the narrower UI windows normally used in vertical split mode. This constraint is used for all relevant UI windows in Combat Simulator, but was missing on a few that don't show up (which I think was: objectives, briefing, results).
- `set_chr_team(CHR_ANTI, TEAM_*)` has been special-cased to set the team of all enemy players. At the very least, it stops enemy players from autoaiming at each other. Levels where this has an effect: Villa, Chicago, Rescue, Air Base, Air Force One.

----

And now for a list of known (and likely) issues!

- This doesn't add another flag for the save file to handle an extra bit for selecting the current player, so if you pick anything other than Player 1 as the main player, then reloading the program will select Player 2. At least I think that's what it picks.
- ~~This expects players to be allocated sequentially. If you have Players 1, 2, and 4, it will assume you only have 2 players.~~ Fixed in a later commit.
- ~~On the other hand... if you have 4 players plugged in, you will have 4 players, even if you only want to do a 2- or 3-player game. I don't have an option to reduce that down yet.~~ Fixed in a later commit.
- 2.x control schemes will probably break. [UPDATE: I've forced it to use the combat simulator way of assigning inputs, but 2.x may still break.]

And here's the per-level issues I've left in because they can be dealt to later and most of them are minor:

- Investigation: Radiation only affects the first of the enemy players, not any of the extra enemy players. (This is the most annoying one to attempt to fix, and will require this level to be special-cased.)
- Chicago: If any extra enemy player dies in front of the building entrance guard, the entrance is sealed. (I actually tried to fix this one, but setting the teams correctly isn't enough for this.)
- *Rescue: This was an issue earlier on, but there was a way for an extra enemy player to destroy the crate which involved the Dragon's secondary proxy mine function. There might be a vanilla bug somewhere. Either way, I hit this bug before inverting the explosion counter-op player logic. (Also, who even uses the crate, anyway?)*
- Deep Sea: The extra enemy players are not marked as invincible during the outro. I'm not sure if this actually has any effect, however - it looks like all players are made invisible and possibly even invincible during cutscenes, but I haven't looked at the code so can't say for sure.
- CI Defense: If the bomb explodes, an explosion will only be created around the main player and the first enemy player. The mission is still eventually explicitly failed.
- Various levels: Certain things are prevented from appearing or disappearing when either the main player, the co-op player, or the first enemy player are in certain locations. Currently, the extra enemy players are not factored in. The cases I'm aware of are:
  - Extraction: Cassandra disappearing (ironically this is bugged in co-op, as from what I can tell, the guard ends up being: `if (!(anti || (bond && coop))) disappear();`)
  - Air Base: Certain characters appearing or disappearing during the takeover
  - CI Defense: The bomb and 6 Mr. Blondes appearing; Hostages unloading
  - Attack Ship: Elvis and Maians appearing... and there's something to do with some "shy Skedar" thing
  - Maian SOS: Characters from the first part of the level being unloaded

Lastly, a bug that isn't ours:

- Attack Ship has some `if_chr_activated_object(CHR_ANTI, ...)` calls which has something to do with some of the lifts in the level, but the actual implementation can only handle the case of the main player and the coop player. So these actually do nothing.

iamgreaser committed:
Preliminary 4-player counter-op mode

Here's some things that will be needed off the top of my head:

- Handling the half-screen-width case for the pause menus
- Changing the player select menu to select the not-counter-op player instead, then extending that to 4 players
- Thorough testing of each level for any weird counter-op-specific features

iamgreaser committed:
Use vertical-split versions of menus in 3+-player main missions

These work fine in the 2x2 split arrangement.

iamgreaser committed:
net: Use `LOCALPLAYERCOUNT()` for UI size determination

iamgreaser committed:
Add macros for counter-op player determination; Handle some missing checks

iamgreaser committed:
Report correct ending success/failure when any counter-op player aborts

iamgreaser committed:
Replace "Counter-Operative" player selection with "Main Player" + extend to 4 entries

iamgreaser committed:
`set_chr_team(CHR_ANTI, *)` now sets all counter-op player teams

The easiest-to-test effect of this is that counter-op players will no longer attempt to auto-aim against one another (p2 vs any other).

Maps where this was necessary: Villa, Chicago, Rescue, Air Base, Air Force One
(Other maps stick to the default of `TEAM_ENEMY`, and so this doesn't really affect it.)

fgsfdsfgs commented:
> This expects players to be allocated sequentially.
There's a player order system in the game that should take care of this (see `g_Vars.playerorder`). Can you leverage that somehow?

> 2.x control schemes will probably break.
What is the issue with that?

NeonNyan commented:
2.* was disabled on multiplayer configs even on vanilla. Dual stick support in Extended mode / Analog walking is handled as an additional stick on each controller.

fgsfdsfgs commented:
Oh, right.

iamgreaser commented:
> There's a player order system in the game that should take care of this (see `g_Vars.playerorder`). Can you leverage that somehow?

I'll look into how that works and get it implemented.

Elaborating on 2.x: There is some special code to assign which controller is the second controller for co-op and counter-op - the code pairs players 1+3 and 2+4 together. I have it set up so if you have more than 2 players, it ignores the special code and goes with the code it uses for Combat Simulator instead.

NeonNyan commented:
Oh, I had no idea it did that.

iamgreaser commented:
I'm not sure if I need the player order system to make this work, and I'm not sure if it's the right thing to use.

I primarily need to find a player which isn't the `bond` player in order to ensure that the `anti` player is set, and then I can potentially use whatever `joyGetConnectedControllers()` spits out and use the mask directly or count the bits, whichever is appropriate.

I'm also somewhat tempted to tell local multiplayer 2.x to go shove it when `MAX_COOPCHRS` isn't 2, though. It should make things easier to reason with. If someone wants true N64 2.x, we'd be better off having a system to allow people to assign 2 controllers to a player.

But I'll have a go at seeing if I can make this work with gaps in the enabled players list.

iamgreaser committed:
Handle cases for Counter-Op where certain local players can be missing; Don't support 2.x controls for Counter-Op

I'm not sure if this happens in practice in the port itself, but with a bit of temporary hackery one can force certain players to be absent.
This handles that case cleanly.

iamgreaser commented:
And updated. I'm not sure how 2.x will behave under these circumstances but feel free to \@ me if such an issue shows up - if it actually happens before we have a clean way of doing 2.x, I'll put some code in to detect 2.x and force it to use 1.x instead.

But anyway, now if there's somehow a gap in the player list, it should still work fine.

fgsfdsfgs commented:
Going to merge this and see what happens then.
